src/DNSC/Cache.hs

```diff
@@ -0,0 +1,328 @@
+import Data.Map (Map)
+import qualified Data.Map.Strict as Map
+
+import Data.PSQueue (Binding ((:->)), PSQ)
+import qualified Data.PSQueue as PSQ
+import Data.IP (IPv4, IPv6)
+import Network.DNS
+  (Domain, CLASS, TTL, TYPE (..), RData (..),
+   ResourceRecord (ResourceRecord), DNSMessage)
+import qualified Network.DNS as DNS
+
+
+---
+
+type CDomain = ShortByteString
+type CMailbox = ShortByteString
+type CTxt = ShortByteString
+
+data CRSet
+  = CR_A [IPv4]
+  | CR_NS [CDomain]
+  | CR_CNAME CDomain
+  | CR_SOA CDomain CMailbox
+    Word32 Word32 Word32 Word32 Word32
+  | CR_PTR [CDomain]
+  | CR_MX [(Word16, CDomain)]
+  | CR_TXT [CTxt]
+  | CR_AAAA [IPv6]
+  deriving (Eq, Ord, Show)
+
+---
+
+-- Ranking data (section 5.4.1 of RFC2181 - Clarifications to the DNS Specification)
+-- <https://datatracker.ietf.org/doc/html/rfc2181#section-5.4.1>
+
+data Ranking_
+{- + Data from a primary zone file, other than glue data, -}
+  --
+{- + Data from a zone transfer, other than glue, -}
+  --
+{- + The authoritative data included in the answer section of an
+     authoritative reply. -}
+  = RankAuthAnswer
+{- + Data from the authority section of an authoritative answer, -}
+  -- -- avoiding issue of authority section in reply with aa flag
+{- + Glue from a primary zone, or glue from a zone transfer, -}
+  --
+{- + Data from the answer section of a non-authoritative answer, and
+     non-authoritative data from the answer section of authoritative
+     answers, -}
+  | RankAnswer
+{- + Additional information from an authoritative answer,
+     Data from the authority section of a non-authoritative answer,
+     Additional information from non-authoritative answers. -}
+  | RankAdditional
+  deriving (Eq, Ord, Show)
+
+type Ranking = Down Ranking_  -- upper rank is better
+
+rankAuthAnswer, rankAnswer, rankAdditional :: Ranking
+rankAuthAnswer  =  Down RankAuthAnswer
+rankAnswer      =  Down RankAnswer
+rankAdditional  =  Down RankAdditional
+
+rankedSection :: Maybe Ranking -> Maybe Ranking -> (DNSMessage -> [ResourceRecord])
+              -> DNSMessage -> Maybe ([ResourceRecord], Ranking)
+rankedSection authRank noauthRank section msg =
+  (,) (section msg)
+  <$> if DNS.authAnswer flags then authRank else noauthRank
+  where
+    flags = DNS.flags $ DNS.header msg
+
+rankedAnswer :: DNSMessage -> Maybe ([ResourceRecord], Ranking)
+rankedAnswer =
+  rankedSection
+  (Just rankAuthAnswer)
+  (Just rankAnswer)
+  DNS.answer
+
+rankedAuthority :: DNSMessage -> Maybe ([ResourceRecord], Ranking)
+rankedAuthority =
+  rankedSection
+  Nothing  -- avoid security hole with authorized reply and authority section case
+  (Just rankAdditional)
+  DNS.authority
+
+rankedAdditional :: DNSMessage -> Maybe ([ResourceRecord], Ranking)
+rankedAdditional =
+  rankedSection
+  (Just rankAdditional)
+  (Just rankAdditional)
+  DNS.additional
+
+---
+
+data Key = K CDomain TYPE CLASS deriving (Eq, Ord, Show)
+data Val = V CRSet Ranking deriving Show
```

Comment:
`K` and `V` are exported. I guess you'd better to choose more meaningful names.

Reply:
https://github.com/khibino/dns-resolver/pull/2/commits/c93432cfe42f7fd2ba61d7c99d0d831eb4757782
https://github.com/khibino/dns-resolver/pull/2/commits/afc58b5a7b3fca74489f8927fb86263f419a0584